### PR TITLE
feat(igxGrid): Update navigation to include ghost rows.

### LINKED ISSUE #7087: Keyboard navigation with Row pinning
## Description  
Test kb nav with row pinning

### PR DESCRIPTION
**Pull request by skrustev**
#6640

Closes #7087

### Additional information (check all that apply):
 - [x] Bug fix
 - [x] New functionality
 - [ ] Documentation
 - [ ] Demos
 - [ ] CI/CD

### Checklist:
 - [x] All relevant tags have been applied to this PR
 - [x] This PR includes unit tests covering all the new code
 - [ ] This PR includes API docs for newly added methods/properties
 - [ ] This PR includes `feature/README.MD` updates for the feature docs
 - [ ] This PR includes general feature table updates in the root `README.MD`
 - [ ] This PR includes `CHANGELOG.MD` updates for newly added functionality
 - [ ] This PR contains breaking changes
 - [ ] This PR includes `ng update` migrations for the breaking changes
 - [ ] This PR includes behavioral changes and the feature specification has been updated with them

### CODE REVIEW

File: projects/igniteui-angular/src/lib/grids/grid-base.directive.ts [hunk 1]
```diff
@@ -5619,10 +5647,14 @@ export class IgxGridBaseDirective extends DisplayDensityBase implements
         if (this.dataView.slice(rowIndex, rowIndex + 1).find(rec => rec.expression || rec.childGridsData)) {
             visibleColIndex = -1;
         }
-        const shouldScrollVertically = this.navigation.shouldPerformVerticalScroll(rowIndex, visibleColIndex);
+        // If the target row is pinned no need to scroll as well.
+        const shouldScrollVertically =
+            !this.isRecordPinnedByIndex(rowIndex) && this.navigation.shouldPerformVerticalScroll(rowIndex, visibleColIndex);
```
Comment (MayaKirova) [inline projects/igniteui-angular/src/lib/grids/grid-base.directive.ts:5652]: As shouldPerformVerticalScroll is called in other places (mrl, hierarchical grid nav services) the check if a record is pinned should be moved inside the method to ensure consistency.

File: projects/igniteui-angular/src/lib/grids/grid-base.directive.ts [hunk 2]
```diff
@@ -5619,10 +5647,14 @@ export class IgxGridBaseDirective extends DisplayDensityBase implements
...
         const shouldScrollHorizontally = this.navigation.shouldPerformHorizontalScroll(visibleColIndex, rowIndex);
         if (shouldScrollVertically) {
-            this.navigation.performVerticalScrollToCell(rowIndex, visibleColIndex,
+            // Only for top pinning we need to subtract pinned count because virtualization indexing doesn't count pinned rows.
+            const scrollRowIndex = this.isRowPinningToTop ? rowIndex - this.pinnedRecordsCount : rowIndex;
+            this.navigation.performVerticalScrollToCell(scrollRowIndex, visibleColIndex,
```
Comment (MayaKirova) [inline projects/igniteui-angular/src/lib/grids/grid-base.directive.ts:5657]: As performVerticalScrollToCell is called in other places (hierarchical grid and mrl nav services) the scroll row index should be resolved inside the method itself to ensure consitency.

**Comment by MayaKirova** [comment]
Base Grid Issue:
When there are rows pinned, navigation does not always scroll the active/selected cell in view.

1) Run the Row Pinning sample
2) Pin the first 3 rows. 
3) Scroll down so that "Cactus Comidas para llevar" is the last visible row (or navigate with arrow down until you reach it).
4) Click the cell and hit arrow down.
Next cell is activated/selected but is not scrolled in view. 
navigation+pinning

File: projects/igniteui-angular/src/lib/grids/grid-base.directive.ts (hunk 2 above)
Comment (MayaKirova) [reply projects/igniteui-angular/src/lib/grids/grid-base.directive.ts:5657]: Also most calculations seem to be using the pinnedRecordsCount, which may return incorrect results if some of the pinned records are filtered (as pinnedRecordsCount is not affected by data operations).

**Comment by MayaKirova** [comment]
HierarchicalGrid issues:

1) When parent grid has pinned rows, children active cell is not scrolled in view..
    1. Run the Row Pinning sample.
    2. Pin 3 rows in hierarchical grid sample for the parent.
    3. Scroll down parent so that bottom of child grid is visible and start navigating up in child.
The active cells are nor scrolled in view.
In case rows are pinned to bottom, navigate down does not scroll cell in view.

Note: This is probably due to the change in the visible view port as previously the whole tbody was used when calculation top/bottom of view port, but now view port is smaller due to the pinned rows, so calculations in _getMaxTop/_getMinBottom in hierarchical nav service should be updated accordingly.

**Comment by ChronosSF** [comment]
@zdrawku , we are done with the testing and review of the PR. Your team can go through it as well as we discussed.

**Comment by ChronosSF** [comment]
@Aleksandyr , #1 is not related to pinning. I logged it separately in https://github.com/IgniteUI/igniteui-angular/issues/7204

**Comment by skrustev** [comment]
@Aleksandyr 2 and 3 we've decided that they are sample related issues because it involves only scenarios with a templated column and will try to fix them in the demos, #4 is added to #7143 since it is not related to the navigation.

**Pull request merged**